src/OrleansServiceBus/Providers/Streams/EventHub/EventHubDataAdapter.cs

```diff
@@ -16,13 +17,49 @@ public struct CachedEventHubMessage
     {
         public Guid StreamGuid;
         public long SequenceNumber;
+        public DateTime EnqueueTimeUtc;
+        public DateTime DequeueTimeUtc;
         public ArraySegment<byte> Segment;
     }
 
     /// <summary>
+    /// Replication of EventHub EventData class, reconstructed from cached data CachedEventHubMessage
+    /// </summary>
+    [Serializable]
+    public class EventHubMessage
+    {
+        public EventHubMessage(ref CachedEventHubMessage cachedMessage)
+        {
+            int readOffset = 0;
+            StreamIdentity = new StreamIdentity(cachedMessage.StreamGuid, SegmentBuilder.ReadNextString(cachedMessage.Segment, ref readOffset));
+            SequenceNumber = cachedMessage.SequenceNumber;
+            EnqueueTimeUtc = cachedMessage.EnqueueTimeUtc;
+            DequeueTimeUtc = cachedMessage.DequeueTimeUtc;
+            Properties = SegmentBuilder.ReadNextBytes(cachedMessage.Segment, ref readOffset).DeserializeProperties();
+            object offsetObj;
+            Offset = Properties.TryGetValue("Offset", out offsetObj)
```

Comment:
Is it reasonable to assume that Properties contains "Offset" and "PartitionKey"? I know they're part of the EventData object model, but I'm not sure they're also in the "Properties" bag (and you haven't added any code for it). It might be safer to write EventData.Offset and EventData.PartitionKey into the segment directly.